generic linux package

pazos opened the pull request:
Relates to https://github.com/koreader/koreader/issues/9268
Requires https://github.com/koreader/koreader-base/pull/1504

Usage: 
- `kodev release linux` for native package (same arch as host)
- `LINUX_ARCH=arm release linux` for armhf on a x86_64 host.
- `LINUX_ARCH=arm64 release linux` for arm64 on a x86_64 host.

It produces a `koreader-linux-$ARCH-$VERSION.tar.xz` archive, where $ARCH follows `uname -m` convention {x86_64, armv7l, aarch64}

To generate the debian package from the generic archive the following command is required

`./platform/linux/do_debian_package.sh path_to.tar.xz`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/9351)
<!-- Reviewable:end -->

Frenzie reviewed (approved):
Sounds good, no comments at a glance (besides the minor formatting ones provided by CI).

pazos commented:
I think most of permission fixes (previously on `do_debian` script, now on makefile) can go on the repo itself. Will give it a try.

akemnade commented:
> After this PR lintian reports the bare minimum of non-fixable "issues":
> 
> Lintian

Well, I doubt if they are all really non-fixable. If I understand things correctly, KOReader was born in a hostile environment (like the kindle system) and therefore could not rely on anything there, so it had to include libraries and some hacks to work with hacky kernel-side implementation of various things. And the deb packages as they are now try to be generic so there is a one package works everywhere. That all is understandable and needed.

On the other hand distributions e.g. want to ensure that if they fix a security issue in openssl, every package using openssl will use the fixed version, so they tend to avoid having required libraries in every package. So if they package it every distribution will carry on their own patch to untangle things. So that would be duplicate work.
So I am wondering whether such a generic linux target could also be a starting point for distributions wanting to package KOReader, so there is an option to rely on installed libraries. Maybe that is nothing for this PR but for some follow-up.

pazos commented:
@akemnade: it started in https://github.com/koreader/koreader/issues/3108#issuecomment-447702137 with a very specific scope: https://github.com/koreader/koreader/issues/3108#issuecomment-450427505

We still keep the same spirit, AFAICT.

NiLuJe commented:
Yeah, unvendoring would be a maintenance nightmare for something none of us would actually actively test ;).

So, the burden is definitely on downstream for those sorts of endeavors.

Frenzie commented:
> Yeah, unvendoring would be a maintenance nightmare for something none of us would actually actively test ;).

As an aside, sometimes I'll just use my Debian system libraries to write the Lua stuff, keeping building the library as an exercise for later. It's certainly convenient. But yeah, not actively tested for sure. ^_^

akemnade commented:
Well, for newcomers like me, you get baffled by the build system, could you dump and share your knowledgte about running KOReader that way in a human + possibly machine readable form ;-) These libkoreader-* stuff are not system libraries, so it involves something more than the lua files. To avoid misunderstandings I am not talking about putting such builds as downloadable release packages. A tarball bringing its own dependencies (also as a step towards building a deb package) is IMHO ok.
I am just investigating what it would take to have KOReader as a first class citizen of Alpine and therefore also on pmOS before tackling that endeavor. Foliate (one of the alternatives) has other problems.

pazos commented:
> These libkoreader-* stuff are not system libraries, so it involves something more than the lua files. 

Nope. Some stuff is coded in c/c++ using the lua C API.
Other stuff is still in lua, but uses LuaJIT ffi and thus it still relies on given dynamic libraries.

You can get a nice picture of all of them in https://github.com/koreader/koreader-base


Look at `thirdparty` for both 3rd party native code as well as 3rd party lua code (ie: luarocks' modules). 
Look at `ffi` for lua code that interfaces with said dynamic libraries
Look at `ffi_cdecl` for the kind of symbols expected on those libraries.
Look at root dir for (our) native code. Again most of it interfaces with libraries. The rest are just loaders for specific platforms.

> To avoid misunderstandings I am not talking about putting such builds as downloadable release packages. A tarball bringing 
 its own dependencies (also as a step towards building a deb package) is IMHO ok.

Lol. **Nope**, **No**, **Nein**, **Non**.

> I am just investigating what it would take to have KOReader as a first class citizen of Alpine and therefore also on pmOS before tackling that endeavor. 

As far as I can tell nobody built KOReader (as you can grasp now it is a bunch of lua code that scripts a lot of native deps) for nothing other than the GNU C library.

NiLuJe commented:
Also, in part because of FFI, the pinned versions of our vendored libs *matter*, which is another stroke against devendoring ;).

I wouldn't necessarily expect musl to be a huge problem (after all, we manage to deal with bionic), but, yeah, stuff *will* probably break nonetheless ;)*.

poire-z commented:
And some of these thirparty libraries are patched by us.
```
$ ls base/thirdparty/|wc -l
56
$ ls base/thirdparty/*/*patch | wc -l
52
```
Some for specific platforms stuff (ie Android), some to disable some dependancies, tweaks or remove some behaviour, or to fix or extend them (ie. lunasvg recently).
And we (at least I) would rather not have to be involved with these 56 upstream thirdparties to have them include our changes/updates/fixes (because there are more fun ways to spend our time) - and it could take years anyway for such updated/fixed versions to reach your preferred distribution...
So, easier with third parties libs shipped - and less maintenance/issues for us.

NiLuJe commented:
> And we (at least I) would rather not have to be involved with these 56 upstream thirdparties to have them include our changes/updates/fixes (because there are more fun ways to spend our time) - and it could take years anyway for such updated/fixed versions to reach your preferred distribution...

And a large part of our patches are not suitable for upstream anyway (because hacks or platform-specific or whatever ;)).

akemnade commented:
> > To avoid misunderstandings I am not talking about putting such builds as downloadable release packages. A tarball bringing
> > its own dependencies (also as a step towards building a deb package) is IMHO ok.
> 
> Lol. **Nope**, **No**, **Nein**, **Non**.
> 
So you are objecting against your own patches?! That comment was meant as a +1 for this PR as it is now.

> > I am just investigating what it would take to have KOReader as a first class citizen of Alpine and therefore also on pmOS before tackling that endeavor.
> 
> As far as I can tell nobody built KOReader (as you can grasp now it is a bunch of lua code that scripts a lot of native deps) for nothing other than the GNU C library.

Well, it is a bit bumpy, yes. That involves fixing building of libraries, that I can easily get via system package management.

Emru1 commented:
Is it coming any time soon?

pazos commented:
Okay, rebased and switched from gzip to xz, since it saves +10MB.

Not sure which changes are required on the build server, @Frenzie?

My proposal would be:

- build two generic linux packages for `x86_64` and `aarch64` (avoid arm32 for the moment, the one we ship now has a very old, probably never used, ABI)
- use those packages to generate two debian packages: `amd64` and `arm64`

In the future, if somebody requires arm32, we can ship `armv7l` as a tar.xz paired with a deb for `armhf`. It should be useful only on a few devices, such as the Raspberry Pi 2 and similar boards.

Frenzie commented:
> Not sure which changes are required on the build server, @Frenzie?

A couple of new entries in https://gitlab.com/koreader/nightly-builds/-/blob/81784c7bb9088985802a1f322eb03e181d0bd68e/.gitlab-ci.yml

Likely nothing else at all.

(We do still need slightly updated Debian packaging images to move entirely off 18.04 but the current ones should still work regardless.)

benoit-pierre commented on Makefile:
```diff
@@ -414,31 +425,44 @@ androidupdate: all
+
+	# fix permissions
+	find $(INSTALL_DIR)/linux -type d -print0 | xargs -0 chmod 755
+	find $(INSTALL_DIR)/linux -executable -type f -print0 | xargs -0 chmod 755
```
`find … -executable` will already include directories:
>  -executable
>      Matches files which are executable and directories which are searchable (in a file name resolution sense) by the current user.

But I would just use `chmod -R u=rwX,og=r …`.

pazos replied:
Could you make a nice suggestion with the recursive chmod instead of find?

Frenzie replied:
I figure that's simply supposed to read like this? But let's await confirmation.
```suggestion
	chmod -R u=rwX,og=r $(INSTALL_DIR)/linux
```

benoit-pierre commented on platform/linux/do_debian_package.sh:
```diff
@@ -0,0 +1,113 @@
+else
+    echo "${COMMAND} not found, unable to build Debian package"
+    exit 1
+fi
+
+
+rm -rf tmp-debian
```
```suggestion
fi

rm -rf tmp-debian
```

benoit-pierre commented on platform/linux/do_debian_package.sh:
```diff
@@ -0,0 +1,113 @@
+
+echo "${CHANGELOG}" | gzip -cn9 > "${CHANGELOG_PATH}"
+chmod 644 "${CHANGELOG_PATH}"
+}
+
+
+if [ -z "${1}" ]; then
```
```suggestion
}

if [ -z "${1}" ]; then
```

benoit-pierre commented on platform/linux/do_debian_package.sh:
```diff
@@ -0,0 +1,113 @@
+
+
+if [ -z "${1}" ]; then
+    echo "${0}: can't find KOReader archive, please specify a path to a KOReader tar.gz"
```
```suggestion
    echo "${0}: can't find KOReader archive, please specify a path to a KOReader tar.gz" 1>&2
```

benoit-pierre commented on Makefile:
```diff
@@ -414,31 +425,43 @@ androidupdate: all
+	find $(INSTALL_DIR)/linux -type d \( -name "test" -o -name ".github" \) -print0 | xargs -0 rm -rf
+
+	# fix permissions
+	find $(INSTALL_DIR)/linux -type d -print0 | xargs -0 chmod 755
+	find $(INSTALL_DIR)/linux -executable -type f -print0 | xargs -0 chmod 755
+	find $(INSTALL_DIR)/linux -type f \( -name "COPYING" -o -name "git-rev" -o -name "*manifest" -o -name "*.cff" -o -name "*.css" -o -name "*.desktop" -o -name "*.json" -o -name "*.html" -o -name "*.lua" -o -name "*.pattern" -o -name "*.png" -o -name "*.otf" -o -name "*.po*" -o -name "*.so*" -o -name "*.svg" -o -name "*.template" -o -name "*.tpl" -o -name "*.ttf" \) -print0 | xargs -0 chmod 644
+	find $(INSTALL_DIR)/linux -type f -name "reader.lua" -print0 | xargs -0 chmod 755
```
I would just replace the whole block with:
```suggestion
	chmod -R u=rwX,og=rX $(INSTALL_DIR)/linux
```

pazos commented:
Ok, applied, some issues related to permissions:

```bash
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libcrypto.so.1.1]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libczmq.so.1]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libdjvulibre.so.21]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libfreetype.so.6]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libfribidi.so.0]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libgif.so.7]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libharfbuzz.so.0]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libk2pdfopt.so.2]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/liblept.so.5]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/liblodepng.so]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libmupdf.so]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libpng16.so.16]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libsharpyuv.so.0]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libsqlite3.so]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libssl.so.1.1]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libtesseract.so.3]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libunibreak.so.6]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libutf8proc.so.3]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libwebp.so.7]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libwebpdemux.so.2]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libz.so.1]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libzmq.so.4]
E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libzstd.so.1]
``` 

```
W: koreader: executable-not-elf-or-script [usr/lib/koreader/common/turbo/platform.lua]
W: koreader: executable-not-elf-or-script [usr/lib/koreader/common/turbo/socket_ffi.lua]
W: koreader: executable-not-elf-or-script [usr/lib/koreader/common/turbo/syscall.lua]
```

and this one I'm not sure about:

`W: koreader: package-contains-broken-symlink-wildcard ../../../../share/fonts/truetype/noto/*.ttf [usr/lib/koreader/fonts/noto/*.ttf]`

maybe I copied something wrong on your link_fonts suggestion change?

benoit-pierre commented:
> Ok, applied, some issues related to permissions:
> 
> ```shell
> E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libcrypto.so.1.1]
> […]
> E: koreader: shared-library-is-executable 0755 [usr/lib/koreader/libs/libzstd.so.1]
> ```

On my Arch Linux system, all the libraries in `/usr/lib` are executable, but in the Ubuntu docker image for Focal, only a select few, so I guess you should remove execute rights on those in `do_debian_package.sh`.

> ```
> W: koreader: executable-not-elf-or-script [usr/lib/koreader/common/turbo/platform.lua]
> W: koreader: executable-not-elf-or-script [usr/lib/koreader/common/turbo/socket_ffi.lua]
> W: koreader: executable-not-elf-or-script [usr/lib/koreader/common/turbo/syscall.lua]
> ```

Looks like those file have the wrong permissions in the original checkout.

> and this one I'm not sure about:
> 
> `W: koreader: package-contains-broken-symlink-wildcard ../../../../share/fonts/truetype/noto/*.ttf [usr/lib/koreader/fonts/noto/*.ttf]`
> 
> maybe I copied something wrong on your link_fonts suggestion change?

Sorry, my bad, try this:
```bash
link_fonts() {
    syspath="../../../../share/fonts/truetype/$(basename "$1")"
    for FILE in "$1"/*.ttf; do
        ln -snf "${syspath}/${FILE##*/}" "${FILE}"
    done
}
```

Frenzie commented:
I believe that on *BSD they're not executable either.

benoit-pierre commented:
I would check for the required tools before doing anything in `do_debian_package.sh`.

Diff:
```diff
 Makefile                            |  5 +--
 platform/linux/do_debian_package.sh | 72 ++++++++++++++++++++-----------------
 2 files changed, 41 insertions(+), 36 deletions(-)

diff --git i/Makefile w/Makefile
index 4d094c33b..bda882029 100644
--- i/Makefile
+++ w/Makefile
@@ -455,10 +455,7 @@ linuxupdate: all
 	find $(INSTALL_DIR)/linux -type d \( -name "test" -o -name ".github" \) -print0 | xargs -0 rm -rf
 
 	# fix permissions
-	find $(INSTALL_DIR)/linux -type d -print0 | xargs -0 chmod 755
-	find $(INSTALL_DIR)/linux -executable -type f -print0 | xargs -0 chmod 755
-	find $(INSTALL_DIR)/linux -type f \( -name "COPYING" -o -name "git-rev" -o -name "*manifest" -o -name "*.cff" -o -name "*.css" -o -name "*.desktop" -o -name "*.json" -o -name "*.html" -o -name "*.lua" -o -name "*.pattern" -o -name "*.png" -o -name "*.otf" -o -name "*.po*" -o -name "*.so*" -o -name "*.svg" -o -name "*.template" -o -name "*.tpl" -o -name "*.ttf" \) -print0 | xargs -0 chmod 644
-	find $(INSTALL_DIR)/linux -type f -name "reader.lua" -print0 | xargs -0 chmod 755
+	chmod -R u=rwX,og=rX $(INSTALL_DIR)/linux
 	XZ_OPT=9 tar -C $(INSTALL_DIR)/linux/usr -cvJf $(INSTALL_DIR)/$(LINUX_PACKAGE) .
 
 	rm -rf $(INSTALL_DIR)/linux
diff --git i/platform/linux/do_debian_package.sh w/platform/linux/do_debian_package.sh
index 8139743e5..4e6c2df2d 100755
--- i/platform/linux/do_debian_package.sh
+++ w/platform/linux/do_debian_package.sh
@@ -9,8 +9,8 @@ command_exists() {
 
 link_fonts() {
     syspath="../../../../share/fonts/truetype/$(basename "$1")"
-    for FILE in *.ttf; do
-        ln -snf "${syspath}/${FILE}" "$1/${FILE}"
+    for FILE in "$1"/*.ttf; do
+        ln -snf "${syspath}/${FILE##*/}" "${FILE}"
     done
 }
 
@@ -25,7 +25,8 @@ uname_to_debian() {
 
 write_changelog() {
     CHANGELOG_PATH="${1}/share/doc/koreader/changelog.Debian.gz"
-    CHANGELOG=$(cat << 'END_HEREDOC'
+    CHANGELOG=$(
+        cat <<'END_HEREDOC'
 koreader (0.1) unstable; urgency=low
 
   * Fixes most lintian errors and warnings
@@ -44,26 +45,34 @@ END_HEREDOC
     chmod 644 "${CHANGELOG_PATH}"
 }
 
-if [ -z "${1}" ]; then
+if ! [ -r "${1}" ]; then
     echo "${0}: can't find KOReader archive, please specify a path to a KOReader tar.gz" 1>&2
     exit 1
-else
-    mkdir -p tmp-debian/usr
-    chmod 0755 tmp-debian/usr
-    tar -xf "${1}" -C tmp-debian/usr
-    ARCH="$(echo "${1}" | cut -d '-' -f3)"
-    VERSION="$(cut -f2 -dv "tmp-debian/usr/lib/koreader/git-rev" | cut -f1,2 -d-)"
-    DEB_ARCH="$(uname_to_debian "${ARCH}")"
 fi
 
-# Run only if dpkg-deb exists
-COMMAND="dpkg-deb"
-if command_exists "${COMMAND}"; then
-    BASE_DIR="tmp-debian"
+# Check for required tools.
+missing_tools=()
+for tool in dpkg-deb fakeroot; do
+    if ! command_exists "${tool}"; then
+        missing_tools+=("${tool}")
+    fi
+done
+if [[ ${#missing_tools[@]} -ne 0 ]]; then
+    echo "${0}: unable to build Debian package, the following tools are missing: ${missing_tools[*]}" 1>&2
+    exit 1
+fi
 
-    # populate debian control file
-    mkdir -p "${BASE_DIR}/DEBIAN"
-    cat  >"${BASE_DIR}/DEBIAN/control" <<EOF
+mkdir -p tmp-debian/usr
+chmod 0755 tmp-debian/usr
+tar -xf "${1}" -C tmp-debian/usr
+ARCH="$(echo "${1}" | cut -d '-' -f3)"
+VERSION="$(cut -f2 -dv "tmp-debian/usr/lib/koreader/git-rev" | cut -f1,2 -d-)"
+DEB_ARCH="$(uname_to_debian "${ARCH}")"
+BASE_DIR="tmp-debian"
+
+# populate debian control file
+mkdir -p "${BASE_DIR}/DEBIAN"
+cat >"${BASE_DIR}/DEBIAN/control" <<EOF
 Section: graphics
 Priority: optional
 Depends: libsdl2-2.0-0, fonts-noto-hinted, fonts-droid-fallback, libc6 (>= 2.2.3)
@@ -80,24 +89,23 @@ Description: Ebook reader application supporting PDF, DjVu, EPUB, FB2 and many m
  It’s available for Kindle, Kobo, PocketBook, Android and desktop Linux.
 EOF
 
-    # use absolute path to luajit in reader.lua
-    sed -i 's,./luajit,/usr/lib/koreader/luajit,' "${BASE_DIR}/usr/lib/koreader/reader.lua"
+# use absolute path to luajit in reader.lua
+sed -i 's,./luajit,/usr/lib/koreader/luajit,' "${BASE_DIR}/usr/lib/koreader/reader.lua"
 
-    # use debian packaged fonts instead of our embedded ones to save a couple of MB.
-    # Note: avoid linking against fonts-noto-cjk-extra, cause it weights ~200MB.
-    link_fonts "${BASE_DIR}/usr/lib/koreader/fonts/noto"
+# use debian packaged fonts instead of our embedded ones to save a couple of MB.
+# Note: avoid linking against fonts-noto-cjk-extra, cause it weights ~200MB.
+link_fonts "${BASE_DIR}/usr/lib/koreader/fonts/noto"
 
-    # DroidSansMono has a restrictive license. Replace it with DroidSansFallback
-    ln -snf ../../../../share/fonts-droid-fallback/truetype/DroidSansFallback.ttf "${BASE_DIR}/usr/lib/koreader/fonts/droid/DroidSansMono.ttf"
+# DroidSansMono has a restrictive license. Replace it with DroidSansFallback
+ln -snf ../../../../share/fonts-droid-fallback/truetype/DroidSansFallback.ttf "${BASE_DIR}/usr/lib/koreader/fonts/droid/DroidSansMono.ttf"
 
-    # add debian changelog
-    write_changelog "${BASE_DIR}/usr"
+# lintian complains if shared libraries have execute rights.
+find "${BASE_DIR}" -type f -perm /+x -name '*.so*' -print0 | xargs -0 chmod a-x
 
-    fakeroot dpkg-deb -b "${BASE_DIR}" "koreader-${VERSION}-${DEB_ARCH}.deb"
-else
-    echo "${COMMAND} not found, unable to build Debian package" 1>&2
-    exit 1
-fi
+# add debian changelog
+write_changelog "${BASE_DIR}/usr"
+
+fakeroot dpkg-deb -b "${BASE_DIR}" "koreader-${VERSION}-${DEB_ARCH}.deb"
 
 rm -rf tmp-debian
 
```

Building the deb on Arch Linux, and running lintian in docker:
```
E: koreader: embedded-library usr/lib/koreader/libs/libfreetype.so.6: freetype
E: koreader: embedded-library usr/lib/koreader/libs/libjpeg.so.8: libjpeg
E: koreader: embedded-library usr/lib/koreader/libs/libmupdf.so: openjpeg
E: koreader: embedded-library ... use --no-tag-display-limit to see all (or pipe to a file/program)
W: koreader: duplicate-font-file usr/lib/koreader/fonts/freefont/FreeSans.ttf also in fonts-freefont-ttf
W: koreader: duplicate-font-file usr/lib/koreader/fonts/freefont/FreeSerif.ttf also in fonts-freefont-ttf
W: koreader: executable-not-elf-or-script usr/lib/koreader/common/turbo/platform.lua
W: koreader: executable-not-elf-or-script usr/lib/koreader/common/turbo/socket_ffi.lua
W: koreader: executable-not-elf-or-script usr/lib/koreader/common/turbo/syscall.lua
W: koreader: opentype-font-wrong-filename usr/lib/koreader/fonts/freefont/FreeSans.ttf
W: koreader: opentype-font-wrong-filename usr/lib/koreader/fonts/freefont/FreeSerif.ttf
W: koreader: script-not-executable usr/lib/koreader/plugins/terminal.koplugin/profile
```

How come our freefont files use the wrong extension?

pazos commented:
thanks for the diff, @benoit-pierre 

> How come our freefont files use the wrong extension?

No idea. Also no idea why I didn't unvendor them. 

I can't remember if our freefont is special :p, pinging @NiLuJe

pazos commented on platform/linux/do_debian_package.sh:
```diff
@@ -75,7 +75,7 @@ mkdir -p "${BASE_DIR}/DEBIAN"
 Section: graphics
 Priority: optional
-Depends: libsdl2-2.0-0, fonts-noto-hinted, fonts-droid-fallback, libc6 (>= 2.2.3)
+Depends: libsdl2-2.0-0, fonts-noto-hinted, fonts-droid-fallback, fonts-freefont-ttf, libc6 (>= 2.2.3)
```
> libc6 (>= 2.2.3)

This was true at some point. Not sure what's the glibc version of the current build machine image. Maybe it needs a bump?

benoit-pierre replied:
```
ii  libc6:amd64    2.31-0ubuntu9.15 amd64        GNU C Library: Shared libraries
```

pazos replied:
good, same in [bullseye](https://packages.debian.org/es/bullseye/libc6)

NiLuJe commented:
> I can't remember if our freefont is special :p, pinging @NiLuJe

Yes, IIRC, I built then from SVN manually.

And I had to mangle the file extension to avoid migration annoyances.

pazos commented:
> > I can't remember if our freefont is special :p, pinging @NiLuJe
> 
> Yes, IIRC, I built then from SVN manually. (Details should be in the commit message for each font build).
> 
> And I had to mangle the file extension to avoid migration annoyances (but they are built as CFF), as the original upstream build we used was a TrueType one.

Thanks!

I'm assuming all the symbols used across the UI come from nerdfonts. If so and our freefont has the same glyphs as the debian package I'm mostly in for unvendor it.

poire-z commented:
I don't know about the Free* fonts - and if it is still active upstream.
But our nerdfont is very special, with twisted glyph mappings, so we should not use an other one.
Another one we'd better keep the shiiped version is Noto Sans CJK sc - we tried in the past more recent ones, and they had issues (with some Japanese dot if I remember right).
Some symbols may come from our main Noto Sans, some from NotoSansCJK (ie. the chapter prefix in Wikipedia lookup results and epubs), a lot from nerdfont - and possibly some rare ones from FreeSans/Serif which have a lot of coverage.

I think we would avoid some bug reports, non-reproducibility, and tough investigations if we were to keep using our shipped fonts. But up to you, this must be a very small user base.

NiLuJe commented:
We may very well rely on some obscure SVN-only features of FreeFont somewhere, because the last release is from ***2012*** (!).

Emru1 commented on platform/linux/do_debian_package.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+!/bin/bash
```
missing shebang

pazos replied:
wops, editor hurry :p

benoit-pierre commented on platform/linux/do_debian_package.sh:
```diff
@@ -0,0 +1,109 @@
+write_changelog() {
+    CHANGELOG_PATH="${1}/share/doc/koreader/changelog.Debian.gz"
+    CHANGELOG=$(
+        cat << 'END_HEREDOC'
```
```suggestion
        cat <<'END_HEREDOC'
```

benoit-pierre commented on platform/linux/do_debian_package.sh:
```diff
@@ -0,0 +1,109 @@
+
+# populate debian control file
+mkdir -p "${BASE_DIR}/DEBIAN"
+cat  >"${BASE_DIR}/DEBIAN/control" <<EOF
```
```suggestion
cat >"${BASE_DIR}/DEBIAN/control" <<EOF
```

pazos commented:
Rebased on top of latest Makefile changes. 

Seem to build fine, tested both the x86_64 and the arm64 builds. Cannot test armhf ATM since it isn't available on default arch repos.

benoit-pierre commented on make/linux.mk:
```diff
@@ -0,0 +1,34 @@
+
+	# fix permissions
+	chmod -R u=rwX,og=rX $(INSTALL_DIR)/linux
+	XZ_OPT=9 tar -C $(INSTALL_DIR)/linux/usr -cvJf $(LINUX_PACKAGE) .
```
Why use `$(INSTALL_DIR)/linux/usr/…` everywhere if we just end up stripping `/usr`?

pazos replied:
Good catch

> why?

For no good reason. :p

Frenzie commented:
@pazos `debian` was removed as a target completely? I had kind of assumed the `debian` target would simply depend on the `linux` target and then execute the do_debian_package script.

pazos commented:
It is even on the commit message: https://github.com/koreader/koreader/commit/2c6808ba78bb7dd38117d57643c86e6101aeb5fa :)

But feel free to adjust as you wish to make things comfortable on your end.

Frenzie commented:
I'll just add the script execution, but I guess it seems a bit hidden for "regular" users. :-)

pazos commented:
> I'll just add the script execution, but I guess it seems a bit hidden for "regular" users. :-)

I think "regular" debian users would benefit more from https://github.com/koreader/koreader/issues/9268 than from building the deb package from source ;)

Frenzie commented:
> I'll just add the script execution, but I guess it seems a bit hidden for "regular" users. :-)

I take this back. The reason kodev exists is not to have to do complicated nonsense like that. This will require some investigation but the old situation was basically perfect. :-/

pazos commented:
>> I'll just add the script execution, but I guess it seems a bit hidden for "regular" users. :-)      

> I take this back. The reason kodev exists is not to have to do complicated nonsense like that. This will require some investigation but the old situation was basically perfect. :-/

I can't understand why.  The old `debian` packages are now generic `linux` packages. Users just need two commands to generate a deb package from source.

~~`./kodev release debian`~~
`./kodev release linux`
`./platform/linux/do_debian_package.sh koreader.tar.xz`

Frenzie commented:
The debian target should simply depend on the linux target and do that automatically.

Now it looks like this, which is harder to read and more error-prone.
```
LINUX_ARCH=arm bash build_release.sh linux && bash koreader/platform/linux/do_debian_package.sh koreader/*.tar.xz
```

Whereas it used to look like this:

```
bash build_release.sh debian-armhf
```

Now you can't simply get rid of or rename the do_debian_package.sh script because it's escaped out of its confines; changing it means you have to update things all over the place, which is annoying. It's an irrelevant implementation detail and leaking that out is not ideal.

For context, build_release.sh is a wrapper around kodev that exists mainly for [this](https://gitlab.com/koreader/nightly-builds/-/blob/6f4f354b17c36b05727ea1ff3746048f31de8d03/build_release.sh#L30-38) reason, which probably indicates a pain point in kodev as well, but that aside.

```sh
if [ "$1" = 'android' ]; then
	./kodev build android 2>&1 > "${BUILD_LOG}"
	./kodev release android 2>&1 > "${BUILD_LOG}"
elif [ "$1" = 'android-arm64' ]; then
	ANDROID_ARCH=arm64 ./kodev build android 2>&1 > "${BUILD_LOG}"
	ANDROID_ARCH=arm64 ./kodev release android 2>&1 > "${BUILD_LOG}"
elif [ "$1" = 'android-x86' ]; then
	ANDROID_ARCH=x86 ./kodev build android 2>&1 > "${BUILD_LOG}"
	ANDROID_ARCH=x86 ./kodev release android 2>&1 > "${BUILD_LOG}"
```

But anyway, it's good enough for now.
